FileCopy/cache.cpp

```diff
@@ -0,0 +1,187 @@
+#include "utils.h"
+
+using namespace C150NETWORK;
+using namespace std;
+
+// bounce :: packet_t -> packet_t
+// Performs incoming packet action and constructs response packet in place.
+void bounce(files_t *fs, cache_t *cache, packet_t *p, long count) {
+    bool is_ack = false;
+    fid_t id = p->hdr.id;
+    seq_t seqno = p->hdr.seqno;
+    payload_u *value = &p->value;
+    switch (p->hdr.type) {
+        case SOS:
+            is_ack = false;
+            break;
+        case ACK:
+            is_ack = true;
+            break;
+        case CHECK_IS_NECESSARY:
+            is_ack = idempotent_checkfile(fs, cache, id, seqno,
+                                          value->check.filename,
+                                          value->check.checksum);
+            break;
+        case KEEP_IT:
+            is_ack = idempotent_savefile(fs, cache, id, seqno);
+            break;
+        case DELETE_IT:
+            is_ack = idempotent_deletefile(fs, cache, id, seqno);
+            break;
+        case PREPARE_FOR_BLOB:
+            is_ack = idempotent_prepareforfile(
+                fs, cache, id, seqno, value->prep.filelength,
+                value->prep.nparts, value->prep.filename);
+            break;
+        case BLOB_SECTION:
+            is_ack = idempotent_storesection(
+                fs, cache, id, seqno, value->section.partno,
+                value->section.offset, p->datalen(), value->section.data);
+            break;
+    }
+    p->hdr.type = is_ack ? ACK : SOS;
+    p->hdr.len = sizeof(p->hdr) + sizeof(p->value.total);
+    p->value.total.totalcount = count;
+}
+
+#define SOS false
+#define ACK true
+
+bool bad_seqno(cache_t *c, int id, int seqno) {
+    return c->entries[id].seqno == -1 || c->entries[id].seqno > seqno;
+}
+
+void cache_free(cache_t *cache) {
+    assert(cache);
+    for (auto it : cache->entries) {
+        free(it.second.recvparts);
+        free(it.second.buffer);
+    };
+}
+
+bool idempotent_prepareforfile(files_t *fs, cache_t *cache, fid_t id,
+                               seq_t seqno, int length, int n_parts,
+                               const char *filename) {
+    assert(fs && cache);
+    if (cache->entries[id].seqno > seqno)
+        return SOS;
+    else if (cache->entries[id].seqno == seqno)
+        return ACK;
+
+    // got a brand new prepare!!
+    if (cache->entries[id].recvparts == nullptr)
+        files_register(fs, id, filename, true);
+
+    cache->entries[id].n_parts = n_parts;
+    // realloc in case we've already tried
+    cache->entries[id].recvparts =
+        (bool *)realloc(cache->entries[id].recvparts, sizeof(bool) * n_parts);
+    cache->entries[id].buflen = length;
+    cache->entries[id].buffer =
+        (uint8_t *)realloc(cache->entries[id].buffer, length);
+    for (int i = 0; i < n_parts; i++) cache->entries[id].recvparts[i] = false;
+
+    // new seqno means we are prepared for file
+    cache->entries[id].seqno = seqno;
+    return ACK;
```

Comment:
We should make a local `entry_t &entry = cache->entries[id]` to make this whole thing cleaner